From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jonas Karlman <jonas@kwiboo.se>
Date: Mon, 3 Aug 2026 19:09:41 +0000
Subject: adc: rockchip-saradc: Add driver data for RK3506

The SARADC dt-bindings fallback for RK3506, rockchip,rk3528-saradc,
instructs the driver to request a 1 MHz clock rate. However, CLK_SARADC
used on RK3506 only supports use of a 1-16 divider so requesting 1 MHz
rate overflows and instead result in a rate of 3 MHz.

Add driver data for rockchip,rk3506-saradc that instead requests a
1.5 MHz clock rate, to avoid clock divider overflow in the RK3506 clock
driver.

Signed-off-by: Austin Lane <vidplace7@gmail.com>
---
 drivers/adc/rockchip-saradc.c | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/drivers/adc/rockchip-saradc.c b/drivers/adc/rockchip-saradc.c
index 111111111111..222222222222 100644
--- a/drivers/adc/rockchip-saradc.c
+++ b/drivers/adc/rockchip-saradc.c
@@ -339,6 +339,14 @@ static const struct rockchip_saradc_data rk3399_saradc_data = {
 	.stop = rockchip_saradc_stop_v1,
 };
 
+static const struct rockchip_saradc_data rk3506_saradc_data = {
+	.num_bits = 10,
+	.num_channels = 4,
+	.clk_rate = 1500000,
+	.channel_data = rockchip_saradc_channel_data_v2,
+	.start_channel = rockchip_saradc_start_channel_v2,
+};
+
 static const struct rockchip_saradc_data rk3528_saradc_data = {
 	.num_bits = 10,
 	.num_channels = 4,
@@ -362,6 +370,8 @@ static const struct udevice_id rockchip_saradc_ids[] = {
 	  .data = (ulong)&rk3066_tsadc_data },
 	{ .compatible = "rockchip,rk3399-saradc",
 	  .data = (ulong)&rk3399_saradc_data },
+	{ .compatible = "rockchip,rk3506-saradc",
+	  .data = (ulong)&rk3506_saradc_data },
 	{ .compatible = "rockchip,rk3528-saradc",
 	  .data = (ulong)&rk3528_saradc_data },
 	{ .compatible = "rockchip,rk3588-saradc",
-- 
Armbian

